Expose schema via Lookup#getSchema.

anwar6953 opened the pull request:
https://issues.cask.co/browse/CDAP-12133

Add the ability to get schema from a dataset's properties in Lookup.
Also add two methods to Lookup, for looking up by `byte[]`, instead of `String`.

anwar6953 committed:
Expose schema via Lookup#getSchema.

anwar6953 committed:
Introduce two methods to Lookup, for lookup by byte[].

archilmedes committed:
Adds changes to tests

archilmedes commented:
Added unit tests, ready for review

anwar6953 commented:
https://builds.cask.co/browse/CDAP-DUT5943-1

anwar6953 committed:
Fix Checkstyle.

anwar6953 committed:
Fix checkstyle.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Lookup.java:
```diff
@@ -16,15 +16,43 @@
  *
- * @param <T> the type of object that will be returned for a lookup
+ * @param <T> the type of object that will be returned for a lookup with String as the key
+ * @param <R> the raw type of object that will be returned for lookups with byte[] as the key
```
This is confusing to me, when would we have a lookup where the value is a different type for different keys?

albertshau replied:
Instead of changing this API, it may be better for LookupProvider to return a Lookup that does the correct value translation based on provided arguments.

anwar6953 replied:
For KeyValueTable lookup, there is a method:
`String get(String key)`, where the key and value are both Strings.
There is also the method:
`byte[] get(String key)`, where the value type is different (both key and value are `byte[]`).

anwar6953 replied:
Regarding your second comment, I don't think that LookupProvider should return an object that does the translation upon the schema, since the caller may want to do a custom transformation from bytes into an object. Not every caller would want a StructuredRecord.

albertshau replied:
The comment wasn't about returning a StructuredRecord, but was about why we need to add this second parameter. The value is already a generic, shouldn't there be a way to get a `Lookup<byte[]>` instead of a `Lookup<String>`? Maybe based on the arguments passed into provide()?

Otherwise it's really strange, I have `Lookup<String, byte[]>` then `lookup("abc")` returns a different type than `lookup(Bytes.toBytes("abc"))`. If anything, I would expect the second parameter to be for the key type.

anwar6953 replied:
The problem is that for TableLookup, the key can be byte[] or String, and the value can be Row or StructuredRecord
The KeyValueTableLookup key can be byte[] or String, and the value can be byte[] or String.

Supporting this is difficult with the current Lookup interface. I'm going to update the Lookup interface to not have all of those methods.

albertshau commented on cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Lookup.java:
```diff
@@ -16,15 +16,43 @@
+   * @param keys the keys to lookup
+   * @return a map from key to value
+   */
+  Map<byte[], R> lookup(byte[]... keys);
```
To be consistent, seems like there should be a method that takes a Set of byte[].

anwar6953 replied:
I couldn't add that, because the method signature would be the same for the added method (simply a different parameter in the Set parameterization doesn't change the method signature).
We could rename the new method, but then it wouldn't be consistent anymore, and I didn't see a strong use case for such a method.

albertshau commented on cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/LookupTransform.java:
```diff
@@ -65,6 +68,29 @@ public void initialize(TransformContext context) throws Exception {
   public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
     T lookedUpValue = lookup.lookup((String) input.get(config.lookupKey));
+    R lookedUpValueBytes = lookup.lookup(((String) input.get(config.lookupKey)).getBytes());
+    if (lookedUpValue instanceof Result) {
```
this is really brittle and ties this class directly to the test case. lookup type and schema should be made into optional config properties.

archilmedes replied:
Improved upon this.

archilmedes committed:
Unties LookupTransform from test cases

albertshau commented on cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/LookupTransform.java:
```diff
@@ -65,6 +68,14 @@ public void initialize(TransformContext context) throws Exception {
     T lookedUpValue = lookup.lookup((String) input.get(config.lookupKey));
+    R lookedUpValueBytes = lookup.lookup(((String) input.get(config.lookupKey)).getBytes());
+    if (lookedUpValue instanceof Result) {
+      assert lookedUpValueBytes instanceof Result;
```
throw an exception instead of using assert. if the jvm is run without a specific flag, assertions won't do anything.

archilmedes replied:
true, I will change this.

albertshau commented on cdap-app-templates/cdap-etl/hydrator-test/src/main/java/co/cask/cdap/etl/mock/batch/LookupTransform.java:
```diff
@@ -81,7 +92,9 @@ public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter)
     Schema outSchema = Schema.recordOf(input.getSchema().getRecordName(), outFields);
-
+    if (lookup.getSchema() != null) {
+      assert lookup.getSchema().equals(outSchema);
```
throw an exception instead of using assert

archilmedes committed:
Replaces asserts with exceptions in LookupTransform

nitinmotgi reviewed (commented):
Does this now allow to do lookups from Spark ? Also would be good if the lookup returns StructuredRecord that has Type and value.

albertshau commented:
No lookups still do not work in Spark

anwar6953 commented:
Putting this on hold; will resume work on it as soon as I finish the implementation for https://issues.cask.co/browse/CDAP-12084.

anwar6953 committed:
wip
Note to self:
what if we added a new method to LookupProvider that looked something like
<T> Lookup<T> provide(String table, Class<T> valueClass, Map<String, String> arguments)

anwar6953 commented:
This JIRA is moved out of 4.3, will reopen after the 4.3 release.
